src/Carbon/Carbon.php

```diff
@@ -2574,6 +2574,46 @@ public function isLastOfMonth()
         return $this->day === $this->daysInMonth;
     }
 
+    /**
+     * Check if the instance is start of day / midnight.
+     *
+     * @return bool
+     */
+    public function isStartOfDay()
+    {
+        return $this->format('H:i:s.u') === $this->copy()->startOfDay()->format('H:i:s.u');
+    }
+
+    /**
+     * Check if the instance is end of day.
+     *
+     * @return bool
+     */
+    public function isEndOfDay()
+    {
+        return $this->format('H:i:s.u') === $this->copy()->endOfDay()->format('H:i:s.u');
```

Comment:
Could fail with object like this in PHP 7.0: `Carbon::create('2000-01-01 20:34:12.123456')->endOfDay()`, it's explicitly end of day. Then copy operation for `isStartOfDay` and `isEndOfDay` seems heavy. I think we can be more flexible by ignoring microseconds part and could simply be: `$this->format('H:i:s') === '00:00:00'` and `$this->format('H:i:s') === '23.59.59'`

Reply:
I was comparing format with string originally, but decided to rely on startOfDay/endOfDay to keep the "is" check always consistent with these methods. Could be overkill, but PHP got so much faster I rarely think about these micro-optimizations anymore.

Date only DateTime (like for MySQL date column) will have zeroed time part, microseconds included. Thus in my opinion dates with microseconds should not be considered date only. But I guess this depends on use case and could serve us better as separate `isDateOnly`, while we can discuss how to deal with microseconds in isStartOfDay/isEndOfDay separately. Should we cut them off or round them?

Reply:
False, I configure my MySQL DateTime to (6) precision, that's means microseconds are stored and returned and it's a best practice. And microseconds zeroed is good for your isStartOfDay but will be always wrong for your isEndOfDay so it's more likely an argument in favor of a floor comparison. The fact is it will not be consistent in PHP 7.0: event `10:10:10`->endOfDay will give: `10:10:10.999999` so you will never get isEndOfDay truthy in PHP 7.0 if the date hadn't be created originally with `.999999` microseconds.

You may also consider this rule is not relevant: `$this->startDate && !$this->startDate->isStartOfDay() || $this->endDate && !$this->endDate->isStartOfDay()` because you can fall on start/end of day by accident. It may be better to always display hour part, then have other `toString` method, like `toDateRangeString`, then think about new `is*` methods in a dedicated issue.

Reply:
MySQL `DATE` not `DATETIME`. Latter obviously contains time.

I started to think about `2018-05-16 00:00:00.000000` as equivalent to `2018-05-16` because of how MySQL treats date-only datetime - as precisely midnight on that day. If you insert `2018-05-16` it will be expanded to `2018-05-16 00:00:00.000000` if you select where `x < 2018-05-16` you'll get only days prior to given. So it doesn't matter if start and end are both midnight by accident - cutting of `0`s is just an equivalent notation, but more compact and easier for humans.

I get that MySQL is not a full picture, I just didn't think about `toString` representation as needing to be so precise. Just like Carbon's `toString` ignores microseconds. It will never be precise, because we cannot know what the filters do. I was thinking about it more as a debug method "uh oh what does this one contain? let's echo it.". Probably should've made it clear before adding translations.

I don't get your PHP 7 example. How come endOfDay will give: 10:10:10.999999? Or do you mean if I do `Carbon::parse('... 23:59:59')` it won't be endOfDay? Well same will be true for `Carbon::parse('... 23:59')`...